compiler/rustc_parse/src/parser/attr_wrapper.rs

```diff
@@ -23,26 +30,181 @@ use rustc_span::{Span, DUMMY_SP};
+// and `num_calls`, we can reconstruct the `TokenStream` seen
+// by the callback. This allows us to avoid producing a `TokenStream`
+// if it is never needed - for example, a captured `macro_rules!`
+// argument that is never passed to a proc macro.
+// In practice token stream creation happens rarely compared to
+// calls to `collect_tokens` (see some statistics in #78736),
+// so we are doing as little up-front work as possible.
+//
+// This also makes `Parser` very cheap to clone, since
+// there is no intermediate collection buffer to clone.
+#[derive(Clone)]
+struct LazyTokenStreamImpl {
+    start_token: (Token, Spacing),
+    cursor_snapshot: TokenCursor,
+    num_calls: usize,
+    break_last_token: bool,
+    replace_ranges: Box<[ReplaceRange]>,
+}
+
+#[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
+rustc_data_structures::static_assert_size!(LazyTokenStreamImpl, 144);
+
+impl CreateTokenStream for LazyTokenStreamImpl {
+    fn create_token_stream(&self) -> PreexpTokenStream {
+        let num_calls = self.num_calls;
+        // The token produced by the final call to `next` or `next_desugared`
+        // was not actually consumed by the callback. The combination
+        // of chaining the initial token and using `take` produces the desired
+        // result - we produce an empty `TokenStream` if no calls were made,
+        // and omit the final token otherwise.
+        let mut cursor_snapshot = self.cursor_snapshot.clone();
+        let tokens = std::iter::once(self.start_token.clone())
+            .chain((0..self.num_calls).map(|_| {
+                if cursor_snapshot.desugar_doc_comments {
+                    cursor_snapshot.next_desugared()
+                } else {
+                    cursor_snapshot.next()
+                }
+            }))
+            .take(num_calls as usize)
+            .map(|(token, spacing)| (FlatToken::Token(token), spacing));
+
+        if !self.replace_ranges.is_empty() {
+            let mut tokens: Vec<_> = tokens.collect();
+
+            let mut replace_ranges = self.replace_ranges.clone();
+            replace_ranges.sort_by_key(|(range, _)| range.start);
+            replace_ranges.reverse();
+
+            for (range, new_tokens) in replace_ranges.iter() {
+                assert!(!range.is_empty(), "Cannot replace an empty range: {:?}", range);
+                // Replace ranges are only allowed to decrease the number of tokens.
+                assert!(
+                    range.len() >= new_tokens.len(),
+                    "Range {:?} has greater len than {:?}",
+                    range,
+                    new_tokens
+                );
+
+                // Replace any removed tokens with `FlatToken::Empty`.
+                // This keeps the total length of `tokens` constant throughout the
+                // replacement process, allowing us to use all of the `ReplaceRanges` entries
+                // without adjusting indices.
```

Comment:
Can the replacement ranges overlap or they are all disjoint?

Reply:
A replace range can be entirely contained in another replace range (partial overlap is not allowed). For example, `#[cfg_attr] struct Foo { #[cfg_attr] field: u8 }` will result in two replace ranges:

1. `#[cfg_attr] field: u8`
2. `#[cfg_attr] struct Foo { #[cfg_attr] field: u8 }`

We will process `#[cfg_attr] field: u8` first, since its start token occurs after the start token of the other replace range. Therefore, the tokens will get transformed like this:

1. `#[cfg_attr] struct Foo { #[cfg_attr] field: u8 }`
2. `#[cfg_attr] struct Foo { {{ #[cfg_attr] field: u8 }} }`
3. `{{ #[cfg_attr] struct Foo { {{ #[cfg_attr] field: u8 }} } }}`

Where `{{ }}` represents the `FlatToken::AttrTarget` that we introduce.

Reply:
Could you add some (debug) asserts to ensure this?

Reply:
I added a debug assertions which checks that replace ranges are either disjoint or nested within each other.